src/programs/azimuthal_average/azimuthal_average.cpp

```diff
@@ -0,0 +1,1534 @@
+#include "../../core/core_headers.h"
+
+class
+	AzimuthalAverage : public MyApp
+{
+
+public:
+	bool DoCalculation();
+	void DoInteractiveUserInput();
+
+private:
+};
+
+void azimuthal_alignment(Image *input_stack, Image *input_stack_times_ctf, int number_of_images, int max_iterations, float unitless_bfactor, bool mask_central_cross, int width_of_vertical_line, int width_of_horizontal_line, float inner_radius_for_peak_search, float outer_radius_for_peak_search, float max_shift_convergence_threshold, float start_angle_for_peak_search, float end_angle_for_peak_search, float rotation_step_size, float max_rotation_convergence_threshold, float pixel_size, int number_of_frames_for_running_average, int savitzy_golay_window_size, int max_threads, float *x_shifts, float *y_shifts, float *psi_angles, float *ctf_sum_of_squares);
+void apply_ctf(Image *current_image, CTF ctf_to_apply, float *ctf_sum_of_squares, bool absolute);
+void divide_by_ctf_sum_of_squares(Image *current_image, float *ctf_sum_f_squares); // const float
+void sum_image_direction(Image *current_image, int dim);
+//void sum_image_direction(Image *current_image, Image *directional_image_sum, int dim);
+void average_rotationally(Image *current_image, Image *current_volume);
+void scale_reference(Image *input_stack, int number_of_images, Image *current_volume, float **ctf_parameters, bool absolute, int max_threads);
+void normalize_image(Image* input_image, float pixel_size, float mask_falloff);
+
+IMPLEMENT_APP(AzimuthalAverage)
+
+// override the DoInteractiveUserInput
+
+void AzimuthalAverage::DoInteractiveUserInput()
+{
+	// intial parameters
+	int new_z_size = 1;
+	std::string text_filename;
+	float pixel_size;
+
+	// ctf parameters
+	float acceleration_voltage;
+	float spherical_aberration;
+	float amplitude_contrast;
+	float defocus_1 = 0.0;
+	float defocus_2 = 0.0;
+	float astigmatism_angle = 0.0;
+	float additional_phase_shift = 0.0;
+	bool input_ctf_values_from_text_file;
+	bool phase_flip_only;
+
+	// alignment expert options
+	int max_iterations = 5;
```

Comment:
Good question and I knew it would come up.

I did try to write a brief comment about what the azimuthal_alignment function does; however, I can certainly expand on it:

> // rotational and translational alignment of tubes
// function similar to unblur_refine_alignment in unblur.cpp with a few differences
// 1. Does not do smoothing, 2. Does not shift in y (vertically), 3. Does not subtract current image from sum

Most of the parameters are similar to the default ones in unblur (note translational parameters are identical and we do not mask central cross and running average should be 1). The rotational parameters and max iterations were obtained by processing a sample set of data that gave a converged aligned, vertically-averaged, CTF-corrected sum, while minimizing the computational burden. Convergence was defined by the sharpness of this sum after each iteration. A couple of iterations were sufficient for the particular data set we used and 5 seemed like a good value that could handle most cases. Of course, more iterations is always better and could improve the reference. A similar heuristic approach was employed for setting the rotational alignment parameters, e.g.,  +/- 5 degrees with a step size of 0.25 degrees. Again, refining the rotation angle could help.